File: IndiaCP_DL_Integration/src/main/java/com/barclays/indiacp/dl/utils/ErrorUtils.java
package com.barclays.indiacp.dl.utils;

import com.barclays.indiacp.model.Error;
import com.barclays.indiacp.model.IndiaCPException;
import java.io.PrintWriter;
import java.io.StringWriter;
import javax.ws.rs.core.Response;

public class ErrorUtils {

    public static Response errorHttpResponse(Object errorCode, String errorMessage, String errorDetails) {
        IndiaCPException ex = new IndiaCPException("Internal DL Integration Layer Exception", Error.SourceEnum.DL_INTEGRATION_LAYER);
        return errorHttpResponse(ex = ex, errorCode = errorCode, errorMessage = errorMessage, errorDetails = errorDetails);
    }

    public static Response errorHttpResponse(Throwable ex, String errorMessage) {
        return errorHttpResponse(ex, errorMessage, errorMessage, ex.getMessage());
    }

    public static Response errorHttpResponse(Throwable ex, Object errorCode, String errorMessage, String errorDetails) {
        if (ex instanceof IndiaCPException)
            return Response.status(Response.Status.INTERNAL_SERVER_ERROR).entity(((IndiaCPException)ex).getError()).build();
        else {
            Error error = new Error();
            error.source(Error.SourceEnum.DL_INTEGRATION_LAYER);
            error.code(errorCode.toString());
            error.message(errorMessage == null ? ex.getMessage() : errorMessage);
            error.details(errorDetails == null ? getStackTrace(ex) : errorDetails);
            return Response.status(Response.Status.INTERNAL_SERVER_ERROR).entity(error).build();
        }
    }

    public static String getStackTrace(Throwable aThrowable) {
        StringWriter result = new StringWriter();
        PrintWriter printWriter = new PrintWriter(result);
        aThrowable.printStackTrace(printWriter);
        return result.toString();
    }

}

File: IndiaCP_Corda/india-cp-cordapp/src/main/kotlin/com/barclays/indiacp/cordapp/utilities/ErrorUtils.kt
package com.barclays.indiacp.cordapp.utilities

import com.barclays.indiacp.model.Error
import com.barclays.indiacp.model.IndiaCPException
import java.io.PrintWriter
import java.io.StringWriter
import javax.ws.rs.core.Response

object ErrorUtils {

    fun errorHttpResponse(ex: Throwable, errorCode: Any = "Unknown", errorMessage: String? = null, errorDetails: String? = null): Response {
        if (ex is IndiaCPException)
            return Response.status(Response.Status.INTERNAL_SERVER_ERROR).entity(ex.getError()).build()
        else {
            val error = Error()
            error.source(Error.SourceEnum.DL_R3CORDA)
            error.code(errorCode.toString())
            error.message(errorMessage ?: ex.message)
            error.details(errorDetails ?: getCustomStackTrace(ex))
            return Response.status(Response.Status.INTERNAL_SERVER_ERROR).entity(error).build()
        }
    }

    fun getStackTrace(aThrowable: Throwable): String {
        val result = StringWriter()
        val printWriter = PrintWriter(result)
        aThrowable.printStackTrace(printWriter)
        return result.toString()
    }

    /**
     * Defines a custom format for the stack trace as String.
     */
    fun getCustomStackTrace(aThrowable: Throwable): String {
        //add the class name and any message passed to constructor
        val result = StringBuilder("BOO-BOO: ")
        result.append(aThrowable.toString())
        val NEW_LINE = System.getProperty("line.separator")
        result.append(NEW_LINE)

        //add each element of the stack trace
        val element: StackTraceElement? = null
        for(element in aThrowable.stackTrace) {
            result.append(element)
            result.append(NEW_LINE)
        }
        return result.toString()
    }
}